packages/timeline-state-resolver/src/integrations/tricaster/timelineStateConverter.ts

```diff
@@ -0,0 +1,155 @@
+import { TimelineState } from 'superfly-timeline'
+import {
+	Mappings,
+	MappingTriCaster,
+	MappingTriCasterType,
+	TSRTimelineObjBase,
+	isTimelineObjTriCasterAudioChannel,
+	isTimelineObjTriCasterDSK,
+	isTimelineObjTriCasterME,
+	isTimelineObjTriCasterMixOutput,
+	MappingTriCasterMixEffect,
+	MappingTriCasterDownStreamKeyer,
+	MappingTriCasterAudioChannel,
+	MappingTriCasterMixOutput,
+	TriCasterAudioChannelName,
+	TriCasterMixEffectName,
+	TriCasterMixOutputName,
+	MappingTriCasterInput,
+	isTimelineObjTriCasterInput,
+	TriCasterInputName,
+} from 'timeline-state-resolver-types'
+import * as _ from 'underscore'
+import { TriCasterState } from './triCasterStateDiffer'
+
+type DeepPartial<T> = { [P in keyof T]?: T[P] extends object ? DeepPartial<T[P]> : T[P] }
+
+export class TimelineStateConverter {
+	private meNames: Set<TriCasterMixEffectName>
+	private inputNames: Set<TriCasterInputName>
+	private audioChannelNames: Set<TriCasterAudioChannelName>
+	private mixOutputNames: Set<TriCasterMixOutputName>
+
+	constructor(
+		private readonly getDefaultState: () => TriCasterState,
+		meNames: TriCasterMixEffectName[],
+		inputNames: TriCasterInputName[],
+		audioChannelNames: TriCasterAudioChannelName[],
+		mixOutputNames: TriCasterMixOutputName[]
+	) {
+		this.meNames = new Set(meNames)
+		this.inputNames = new Set(inputNames)
+		this.audioChannelNames = new Set(audioChannelNames)
+		this.mixOutputNames = new Set(mixOutputNames)
+	}
+
+	getTriCasterStateFromTimelineState(
+		timelineState: TimelineState,
+		newMappings: Mappings,
+		deviceId: string
+	): TriCasterState {
+		const resultState = this.getDefaultState()
+		const sortedLayers = this.sortLayers(timelineState)
+
+		_.each(sortedLayers, ({ tlObject, layerName }) => {
+			const mapping = newMappings[layerName] as MappingTriCaster | undefined
+			if (!mapping || mapping.deviceId !== deviceId) {
+				return
+			}
+			switch (mapping.mappingType) {
+				case MappingTriCasterType.MixEffect:
+					this.applyMixEffectState(resultState, tlObject, mapping)
```

Comment:
I guess this is the TSR way of doing it; but it would be more clear if we compute and return the state.

Reply:
The properties have to be applied in place since it would be inefficient to do a deep copy every time, and return a new object being the result. On the other hand if those methods were to do the changes in place but return the resulting object, there still would be no point in assigning it again. Because nothing is async in this class, the result state could become a class member, kind of like [here](https://github.com/tv2/tv-automation-state-timeline-resolver/pull/46#discussion_r1084885547), but that's no good either

Reply:
If the data that we are operating on here is large enough, then the in-place merging can be justified.
The in-place operations can be kept, while improving the clarity of the execution flow. `reduce` could be used like so:

```typescript
  getTriCasterStateFromTimelineState(
    timelineState: TimelineState,
    newMappings: Mappings,
    deviceId: string
  ): TriCasterState {
    const initialResultState = this.getDefaultState()
    const sortedLayers = this.sortLayers(timelineState)

    return Object.values(sortedLayers).reduce((resultState, { tlObject, layerName }) => {
      const mapping = newMappings[layerName] as MappingTricaster | undefined
      if (!mapping || mapping.deviceId !== deviceid) {
        return resultState
      }

      switch (mapping.mappingType) {
        case MappingTriCasterType.MixEffect:
          return this.applyMixEffectState(resultState, tlObject, mapping)
        case MappingTriCasterType.DownStreamKeyer:
          return this.applyDskState(resultState, tlObject, mapping)
        // rest of cases....
      }
    }, initialResultState)
  }
```